### PR TITLE
Refactor wearable apps for best practices of the async/await pattern

### PR DESCRIPTION
**Pull request by swift-kim**
I'm currently investigating common use cases of async/await in Tizen applications, to identify their common misuse patterns and guide application developers with the right principles of writing asynchronous code.

I analyzed the following 5 wearable samples and made some changes when it was possible.

- Alarm
- Calculator
- HeartRateMonitor
- VoiceMemo
- Weather

The common abuse of async patterns includes:

- Calling `async void` from synchronous code in a 'fire-and-forget' manner
  - e.g. Calling `async void Init()` from a class constructor for asynchronous initialization
  - Providing an `async void` method as an action parameter for `Xamarin.Forms.Command`
  - Not awaiting the caller method of `await Navigation.PushAsync()`
- Misuse of threading
  - Overuse of `Task.Run()` or `Task.Factory.StartNew()` for non-CPU-bound code
  - Underuse of threading for CPU-bound code during app startup
- Mixed use of sync and async code
  - All `.Result` and `.Wait()`'s

I tried not to introduce additional logic (e.g. exception handling) and not to affect the original logic as much as possible. The exception is Weather application where there has been a massive amount of refactoring.

Here is the basic async coding guide by Miscrosoft for beginners: https://docs.microsoft.com/en-us/dotnet/csharp/async

If you know any other patterns used by other applications, please share and help me write a rich guide. There are likely missing many important details because most of the applications listed above are single-threaded.

### CODE REVIEW

File: Wearable/Alarm/Alarm/Implements/AlarmPersistentHandler.cs [hunk 1]
```diff
@@ -62,41 +60,28 @@ public static class AlarmPersistentHandler
         /// <param name="properties">AlarmRecordDictionary object to serialize</param>
         /// <seealso cref="IDictionary<string, AlarmRecord>">
         /// <returns> Task return type </returns>
-        public static Task SerializeAlarmRecordAsync(IDictionary<string, AlarmRecord> properties)
+        public static void SerializeAlarmRecord(IDictionary<string, AlarmRecord> properties)
```
Comment (jkpu) [inline Wearable/Alarm/Alarm/Implements/AlarmPersistentHandler.cs:63]: Why did you change task.await() method to sync method. I think this method should be a task for running other thread 
because of saving record take a long time.

File: Wearable/Alarm/Alarm/Views/SavePopupPage.xaml.cs [hunk 2]
```diff
@@ -46,7 +46,13 @@ public SavePopupPage(AlarmRecord record)
             NavigationPage.SetHasNavigationBar(this, false);
             _record = record;
             SetPopupLabel();
-            ClosePopup();
+
+            Appearing += async (s, e) =>
+            {
+                // Close the pop-up after 2 seconds.
+                await Task.Delay(2000);
+                await Navigation.PopToRootAsync();
```
Comment (myroot) [inline Wearable/Alarm/Alarm/Views/SavePopupPage.xaml.cs:54]:
```suggestion
                _ = Navigation.PopToRootAsync();
```

File: Wearable/Alarm/Alarm/ViewModels/AlertPageModel.cs [hunk 3]
```diff
@@ -72,9 +71,9 @@ public AlertPageModel(AlarmRecord record = null)
             Record = record;
         }
 
-        public void StartAlert()
+        public async Task StartAlert()
```
Comment (jkpu) [inline Wearable/Alarm/Alarm/ViewModels/AlertPageModel.cs:74]: There is no need to wait StartAlert() in caller. So I think previous code is more proper code.

File: Wearable/Alarm/Alarm/Implements/AlarmPersistentHandler.cs [hunk 4]
```diff
@@ -62,41 +60,28 @@ public static class AlarmPersistentHandler
...
+            try
             {
-                Stream alarmFileStream = null;
-                var fileOpener = new FileOpener();
-                try
+                using (var alarmFileStream = fileOpener.OpenFile(AlarmRecordStoreFile, FileMode.Create))
                 {
-                    alarmFileStream = fileOpener.OpenFile(AlarmRecordStoreFile, System.IO.FileMode.Create);
-                    using (XmlDictionaryWriter alarmWriter = XmlDictionaryWriter.CreateBinaryWriter(alarmFileStream))
+                    using (var alarmWriter = XmlDictionaryWriter.CreateBinaryWriter(alarmFileStream))
```
[the remark this answers is not in the document]
Comment (myroot) [reply Wearable/Alarm/Alarm/Implements/AlarmPersistentHandler.cs:73]: https://docs.microsoft.com/ko-kr/dotnet/standard/io/asynchronous-file-i-o

File: Wearable/Alarm/Alarm/ViewModels/AlertPageModel.cs [hunk 5]
```diff
@@ -72,9 +71,9 @@ public AlertPageModel(AlarmRecord record = null)
...
         {
-            AlarmNativeHandler.PlaySound();
+            await AlarmNativeHandler.PlaySound();
```
Comment (jkpu) [inline Wearable/Alarm/Alarm/ViewModels/AlertPageModel.cs:76]: Caller have no reason to wait PlaySound() method. so remove `await`

File: Wearable/Alarm/Alarm/Views/AlarmAlertPage.xaml.cs [hunk 6]
```diff
@@ -62,12 +62,12 @@ public AlarmAlertPage(AlertPageModel viewModel)
...
         /// <param name="sender">object</param>
         /// <param name="args">EventArgs</param>
-        void OnDismissButtonClicked(object sender, EventArgs args)
+        async void OnDismissButtonClicked(object sender, EventArgs args)
         {
             Console.WriteLine("OnDismissButtonClicked()");
 
             _viewModel.Dismiss();
-            PopAsyncAlertPage();
+            await Navigation.PopAsync();
```
Comment (jkpu) [inline Wearable/Alarm/Alarm/Views/AlarmAlertPage.xaml.cs:70]: remove await and async this event handler.

File: Wearable/Alarm/Alarm/Implements/AlarmNativeHandler.cs [hunk 7]
```diff
@@ -122,7 +123,7 @@ public static void StopVibration()
         /// <summary>
         /// Asynchronously plays ringing sound
         /// </summary>
-        async public static void PlaySound()
+        public static async Task PlaySound()
```
Comment (myroot) [inline Wearable/Alarm/Alarm/Implements/AlarmNativeHandler.cs:126]: I think, no need to return `Task`, Caller do not interesting end of PlaySound, 
And return value is ambiguous, {end of Playback or end of loading}

File: Wearable/Alarm/Alarm/Alarm.cs [hunk 8]
```diff
@@ -53,21 +53,19 @@ protected override void OnPause()
...
             base.OnResume();
-
         }
 
         /// <summary>
         /// Called when this app control event received.
         /// </summary>
         /// <param name="e">AppControlReceivedEventArgs</param>
-        async protected override void OnAppControlReceived(AppControlReceivedEventArgs e)
+        protected override async void OnAppControlReceived(AppControlReceivedEventArgs e)
```
Comment (myroot) [inline Wearable/Alarm/Alarm/Alarm.cs:68]: I think remove `async` and fix 101 line like below
``` c#
                                _ = currentPage.Navigation.PushAsync(new AlarmAlertPage(_alertPageModel))
```

File: Wearable/HeartRateMonitor/src/HeartRateMonitor/Controls/DialogOK.cs [hunk 9]
```diff
@@ -95,18 +95,16 @@ public string Message
         /// </summary>
         public DialogOK()
         {
-            ExecuteCommand = new Command(Display);
+            ExecuteCommand = new Command(async () => await Display());
```
[the remark this answers is not in the document]
Comment (swift-kim) [reply Wearable/HeartRateMonitor/src/HeartRateMonitor/Controls/DialogOK.cs:98]: As we have discussed offline, if there's a proper reference about async Commands, I think we can allow the original code pattern as an exception similarly to event handlers.

File: Wearable/Alarm/Alarm/Implements/AlarmPersistentHandler.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (swift-kim) [reply Wearable/Alarm/Alarm/Implements/AlarmPersistentHandler.cs:73]: https://github.com/Samsung/Tizen-CSharp-Samples/blob/75b44720c61ecbc83c21ef0f6bc0525e4da692dc/Wearable/Alarm/Alarm/Models/AlarmModel.cs#L204-L207

It's used only synchronously so the behavior is identical. I agree that it should be async but the code chaining should be re-designed for it.

File: Wearable/Alarm/Alarm/Views/SavePopupPage.xaml.cs (hunk 2 above)
Comment (swift-kim) [reply Wearable/Alarm/Alarm/Views/SavePopupPage.xaml.cs:54]: If the user doesn't really (intentionally) care about an error thrown by `Navigation.PopToRootAsync()`, it's also an acceptable pattern.

File: Wearable/Alarm/Alarm/Implements/AlarmNativeHandler.cs (hunk 7 above)
Comment (swift-kim) [reply Wearable/Alarm/Alarm/Implements/AlarmNativeHandler.cs:126]: It's not about the result of the awaitable Task. The concern about async void is exception handling semantics and code execution controllability.

File: Wearable/Alarm/Alarm/ViewModels/AlertPageModel.cs (hunk 3 above)
Comment (swift-kim) [reply Wearable/Alarm/Alarm/ViewModels/AlertPageModel.cs:74]: Although it's a very exceptional case, if an error occurs during `player.PrepareAsync()`, it should be propagated properly. If you don't really care or the exception is already handled somewhere, it could also be `_ = StartAlert()` as @myroot has suggested above.

File: Wearable/Alarm/Alarm/ViewModels/AlertPageModel.cs (hunk 5 above)
Comment (swift-kim) [reply Wearable/Alarm/Alarm/ViewModels/AlertPageModel.cs:76]: Please see the previous comment.

File: Wearable/Alarm/Alarm/Alarm.cs (hunk 8 above)
Comment (swift-kim) [reply Wearable/Alarm/Alarm/Alarm.cs:68]: The same as https://github.com/Samsung/Tizen-CSharp-Samples/pull/161#discussion_r439210251.

**Comment by swift-kim** [comment]
- How to implement fire-and-forget async Commands more elegantly:
  - https://johnthiriet.com/mvvm-going-async-with-async-command
- Why not async void:
  - https://docs.microsoft.com/en-us/dotnet/csharp/async
  - https://docs.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming
  - https://johnthiriet.com/removing-async-void
- Some practical async usage:
  - https://blog.stephencleary.com/2013/01/async-oop-0-introduction.html
- About SynchronizationContext and ConfigureAwait for library code
  - https://medium.com/bynder-tech/c-why-you-should-use-configureawait-false-in-your-library-code-d7837dce3d7f
  - https://blog.stephencleary.com/2012/07/dont-block-on-async-code.html
  - https://devblogs.microsoft.com/dotnet/configureawait-faq
  - https://michaelscodingspot.com/c-deadlocks-in-depth-part-2
- For who are interested:
  - https://blog.stephencleary.com/2013/11/there-is-no-thread.html

File: Wearable/HeartRateMonitor/src/HeartRateMonitor/Controls/DialogOK.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (swift-kim) [reply Wearable/HeartRateMonitor/src/HeartRateMonitor/Controls/DialogOK.cs:98]: Maybe you're right. The modification introduces an [implicit async void delegate](https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#implicit-async-void-delegates) which should be avoided. I should consider switching to an [AsyncCommand](https://johnthiriet.com/mvvm-going-async-with-async-command) implementation.

**Comment by swift-kim** [comment]
Updated with force-push.

- Handle some async operations as fire-and-forget
  - Not very good but acceptable
  - Exception handling (try-catch) could be added
- Allow async void for Commands
  - Treat just as event handlers
  - Consider adding a custom AsyncCommand or 3rd-party async MVVM implementation for better exception handling

**Comment by swift-kim** [comment]
It seems the implementation of NotificationTask in the Weather application was taken from [this MSDN article](https://docs.microsoft.com/en-us/archive/msdn-magazine/2014/march/async-programming-patterns-for-asynchronous-mvvm-applications-data-binding). Actually NotificationTask (NotifyTaskCompletion) is not very useful here because the xaml attributes only bind to `.Result` properties (but not `.ErrorMessage` or else) and such bindings will simply work without NotifyTaskCompletion.

I didn't make significant changes and only removed some redundant code though.

**Comment by swift-kim** [comment]
If there's no more review, the change is ready for merge.
